File: 1282.group-the-people-given-the-group-size-they-belong-to.ts
/*
 * @lc app=leetcode id=1282 lang=typescript
 *
 * [1282] Group the People Given the Group Size They Belong To
 */

// @lc code=start
function groupThePeople(groupSizes: number[]): number[][] {
  let result: number[][] = [];
  let tmp = {};

  for (let i = 0; i < groupSizes.length; i++) {
    let arr = tmp[groupSizes[i]] || [];
    arr.push(i);
    if (arr.length == groupSizes[i]) {
      result.push(arr);
      tmp[groupSizes[i]] = [];
    } else {
      tmp[groupSizes[i]] = arr;
    }
  }

  return result;
}
// function groupThePeople(groupSizes: number[]): number[][] {
//   let result: number[][] = [];
//   let tmp: number[][] = [[]];
//
//   for (let i = 0; i < groupSizes.length; i++) {
//     let arr = tmp[groupSizes[i]] || [];
//     arr.push(i);
//     if (arr.length == groupSizes[i]) {
//       result.push(arr);
//       tmp[groupSizes[i]] = [];
//     } else {
//       tmp[groupSizes[i]] = arr;
//     }
//   }
//
//   return result;
// }
// @lc code=end

File: 1282.group-the-people-given-the-group-size-they-belong-to.rb
#
# @lc app=leetcode id=1282 lang=ruby
#
# [1282] Group the People Given the Group Size They Belong To
#

# @lc code=start
# @param {Integer[]} group_sizes
# @return {Integer[][]}
def group_the_people(group_sizes)
  hash = {}
  result = []

  group_sizes.each_with_index do |v, k|
    hash[v] ||= []
    hash[v].push(k)
    if hash[v].length == v
      result.push(hash[v])
      hash[v] = nil
    end
  end

  result
end
# @lc code=end
